### PR TITLE
Add 'bitflags' modifier for enum.

### LINKED ISSUE #42521: Modifier for bitflag enums
Consider the following code

```ts
export enum ScriptKind {
  Unknown = 0,
  JS = 1,
  JSX = 2,
  TS = 3,
  TSX = 4,
  External = 5,
  JSON = 6,
}
```

If a user tries to pass along the expression `ScriptKind.TS | ScriptKind.TSX` to a parameter that expects a `ScriptKind`, then will end up constructing a value equal to `ScriptKind.JSON`.

Really, the user should have been told that this enum wasn't intended to be used in bitwise operations! But unfortunately, TypeScript doesn't catch this because you can assign any number to any numeric enum.

One could imagine a mode that prevents both the construction and the assignment for certain enums. In this mode, an enum has to be declared with the `bitflags` modifier.

```
bitflags enum SymbolFlags {
    // ...
}
```

Alternatively, this could be something more akin to `/** @deprecated */` or `/** @internal */`.

```
/// @bitflags
enum SymbolFlags {
    // ...
}
```

This is in some part inspired by Rust's [`bitflags`](https://docs.rs/bitflags/1.2.1/bitflags/) macro.

### PR DESCRIPTION
**Pull request by ShuiRuTian**
Fixes #42521

 - [x] Add modifier `bitflags`.
 - [x] bit operator(`|`,`|=`,`&`,`&=`,`^`,`^=`) is only allowed for bitflags enum, if left or right is enum.
 - [x] Add new flag to enable `bitflags` check.

The second one is a big break! Maybe it should be loosen, but how?

And I think there are more features could be added!
- [x] bitflags enum member could only be initialized by BitwiseOperator(&,|,^) or ShiftOperator(<<,>>,>>>)  or bit number
```
bitflags enum BitEnum {
    TS = 1,     // 1 is 1 << 0, OK
    TSX = 2,    // 2 is 1 << 1, OK
    ERR2 = "1", // "1" is not number, error
    All = TS | TSX, // initialized through `|`, should be OK without "1"
    All2 = 3,    // 3 is not a bit, error
    All3 = 1+2 // '+' is not  BitwiseOperator or ShiftOperator, error.
}
```

- [ ] more accurate type.
``` ts
bitflags enum BitEnum {
    TS = 1,
    TSX = 2,
    TSAlias = TS,
    All = TS | TSX,
}
const foo1 = BitEnum.TSAlias;
const foo2 = BitEnum.TSX;
const foo3 = foo1 | foo2;    // what is foo1 type? `BitEnum.TS` or `BitEnum.TSAlias`?
``` 

- [ ] give accurate type when both bitflags enum are known.
``` ts
bitflags enum BitEnum {
    TS = 1,
    TSX = 2,
    All = TS | TSX,
}

var foo1 = BitEnum.TS;
var foo2 = BitEnum.TSX;
var foo3 = foo1 | foo2; // foo3 is possiable to be BitEnum.All?
declare var foo4: BitEnum;
var foo5 = foo4 & foo1; // foo5 is possiable to be BitEnum.TS?

```

---
Some Questions:
1. What should be the flag name? Should it be "strict"? Personally, I want to enable this feature in any new project, but it might be a big break for current project. Now it is `bitEnum`
2. Is there a better name rather than `bitflags`?
3. How to mark enum as `bit`? Should `bitflags` be a modifier or tag or something else?
4. Are error messages easy to understand? I am clearly not a English native speaker, they might looks wired now.

### CODE REVIEW

**Comment by weswigham** [comment]
@ShuiRuTian we were discussing allowing the modifier/tag/whatever in any mode, but only having it do anything different from normal enums today under a new flag - so you can opt in to strictly checking bitflagginess. This way it wouldn't be a straight break for anyone currently using normal enums as bitflags.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -9574,29 +9576,20 @@ namespace ts {
             if (links.declaredType) {
                 return links.declaredType;
             }
-            if (getEnumKind(symbol) === EnumKind.Literal) {
```
Comment (ShuiRuTian) [inline src/compiler/checker.ts:9577]: What is EnumKind on earth?

**Comment by ShuiRuTian** [comment]
@weswigham Thanks a lot! the beta version is ready for review.

But, I find something strange about the enums, which blocks me on `more accurate type.` item in the check list.

And, it is a little confusing for me, not sure whether they are carefully designed or just a bug.

It is basically about `EnumKind`. I feel `EnumKind.Numeric` is something like number, and `EnumKind.Literal` is like string. 

But at the same time, we could see in function `getDeclaredTypeOfEnum`, in the block we have also `getLiteralType` and `getRegularTypeOfLiteralType`. `Literal` for these two functions is more like `Syntax.stringLiteral` and `Syntax.numericLiteral`. 

Also,  it seems if type is EnumKind.Numeric, then enum members does not have any change to have a correct declaredType -- if it is not initialized, in the second time it could only get its parent cached type(Not sure about this, but I am sure enum member in EnumKind.Numeric does not have as rich flags as those members in EnumKind.Literal).

--- 
PS: I change a big else-if to 
```
switch(true){
case (condition):
...
}
```
I think this help debug, the code would jump to the first true condition, pass a lot false condition.
But is this good or at least acceptable? Is this too tricky to make code hard to read?

**Comment by weswigham** [comment]
> I feel EnumKind.Numeric is something like number, and EnumKind.Literal is like string

Numeric is the old bitflag allowing enum, while `Literal` is the newer union-of-uniqe-literal-subtypes enum. You get the later if the enum contains a string initializer today.

**Comment by ShuiRuTian** [comment]
@weswigham thanks for the detailed response. And I meet another case:
```
enum E {
    a = 10,
    b = 10 + 1,
    c = b,
    d = (c) + 1,
    e = 10 << 2 * 8,
}

var q = E.c;
var w = q;
```

When hover on the last `q`, should it shows `var q: E.b`? Or var q: E.c`? Or just `var q: E`?

In this PR, I could make it show as `var q: E.b`.

This is not related to `bitflags` directly, but it is helpful for the last check item feature. And I think it's  not bad to have in all enum.

Is this acceptable? Or we should do it step by step, give a bitflags keyword and basic function first, then more features?

**Comment by ShuiRuTian** [comment]
Seems a new keyword is not acceptable. Would restart working on this if there is any new thought.

**Comment by sandersn** [comment]
To help with PR housekeeping, I'm going to close this draft PR. It is pretty old now.

**Pull request closed without merging**